File: src/test/python/parser-opentrackIO/opentrackIO_parser.py
#!/usr/bin/env python3
#
# opentrackIO_parser.py
#
# Reference code for decoding opentrack messages
# Copyright (c) 2024 Steve Rosenbluth, RiS OSVP camera tracking committee
#
# License: this code is open-source under the FreeBSD License
#
# example: python3 src/test/python/parser-opentrackIO/opentrackIO_parser.py --file=complete_dynamic_example_20240907.json --schema=opentrackio_schema_modified_20240907.json
# exemple: python3 src/test/python/parser-opentrackIO/opentrackIO_parser.py --file=complete_static_example_20240907.json --schema=opentrackio_schema_modified_20240907.json

import json
import argparse
import os

# Class to decode and interpret the protocol
# Pass in either a JSON message or a test filename
class OTProtocol:
    def __init__(self, msg_str=None, filepath=None):
        # msg_str = string containing a single json "sample"
        # filepath: file containing a JSON "sample", for testing
        self.sample_str = msg_str
        self.filepath = filepath
        self.schemapath = None
        self.trans_mult = 1.0                               # user-preferred units multiplier
        self.curtime = 0
        self.cam_trans = {"X":0,"Y":0,"Z":0}
        self.pd = None          # the parsed procotol dictionary
        self.sd = None          # the parsed schema dictionary
        self.sample_time_format = "seconds"
        self.sample_time_formats = ["seconds","milliseconds","microseconds","nanoseconds"]
        self.verbose = False
        self.focus_dist_mult = 1.0

        parser = argparse.ArgumentParser(description='OpenTrackingProtocol parser')
        parser.add_argument('-f', '--file', help='The JSON input file.', default=None)
        parser.add_argument('-s', '--schema', help='The schema (JSON) input file.', default=None)
        parser.add_argument('-t', '--test', help='If present, will only verify json',action='store_true')
        parser.add_argument('-v', '--verbose', help='Make script more verbose',action='store_true')
        args = parser.parse_args()
        if (args.schema):
            if os.path.exists(args.schema):
                self.schemapath = args.schema
        if self.schemapath:
            self.Import_schema()
        if (args.file):
            if os.path.exists(args.file):
                self.filepath = args.file
        if (args.verbose):
            self.verbose = True

        self.Parse()                # parse the actual JSON of the protocol

    def Import_schema(self):
        text = ''
        try:
            if (self.schemapath != None):        
                with open(self.schemapath, 'r') as fd:
                    print("Reading OTIO schema file: {0}".format(self.schemapath))
                    lines = fd.readlines()
                    for line in lines:
                        text = text + line
            self.sd = json.loads(text)
        except json.decoder.JSONDecodeError:
            print("Got JSONDecodeError decoding the sample!")
            self.sd = None
        if not self.sd:
            print("Parse(): Failed to parse OTIO schema file.")
        else:                                   # we have a OTP message
            print("Parsed the schema JSON successfully.")
            if self.verbose:
                print("Contents of the parsed JSON schema dict:\n")
                print(self.sd)
                print("\n")
            #print("camera is: {}".format(self.sd["properties"]["camera"]))
            #print("finding: {}".format(self.sd["properties"]["lens"]["properties"]["focalLength"]["units"]))

    # ingest the text and store the JSON items in a dictionary
    def Parse(self):        
        protocol = None
        text = ''
        if (self.filepath != None):        
            with open(self.filepath, 'r') as fd:
                print("Reading OTIO sample file: {0}".format(self.filepath))
                lines = fd.readlines()
                for line in lines:
                    text = text + line
        elif self.sample_str != None:        
            print("Parsing JSON string from sample buffer...")
            text = self.sample_str
        #print(text)
        #print('\n')    
        try:
            self.pd = json.loads(text)
        except:
            print("Error: failed to parse JSON file! Either path incorrect or JSON invalid.")
            exit(-1)
        if not self.pd:
            print("Parse(): Failed to parse OTIO message file.")
        else:                                   # we have a OTP message
            print("Parsed the sample OTIO JSON successfully.")
            if self.verbose:
                print("Contents of the parsed JSON dict:\n")
                print(self.pd)
                print("\n")

    # Returns the multiplicative conversion factor from meters to user-specified units
    # Valid unit specifier strings: "m", "cm", "mm", "in"
    def Conversion_factor_from_meters(self,unit_str):
        if unit_str == "m":
            return 1.0
        elif unit_str == "cm":
            return 100.0
        elif unit_str == "mm":
            return 1000.0
        elif unit_str == "in":
            return (1000/25.4)

    def Get_camera_trans(self, dimension,object_name=None):
        for tr in self.pd["transforms"]:
            if ("Camera" in tr["name"]):
                if (dimension == 'X'):
                    return tr["translation"]["x"] * self.trans_mult
                elif (dimension == 'Y'):
                    return tr["translation"]["y"] * self.trans_mult
                elif (dimension == 'Z'):
                    return tr["translation"]["z"] * self.trans_mult
                break

    def Get_camera_translations(self,object_name=None):
        return (self.Get_trans('X'), self.Get_trans('Y'), self.Get_trans('Z'))

    def Get_timecode(self):
        if ("timing" in self.pd) and (self.pd["timing"]["timecode"]):
            hh = '{:02}'.format(int(self.pd["timing"]["timecode"]["hours"]))
            mm = '{:02}'.format(int(self.pd["timing"]["timecode"]["minutes"]))                
            ss = '{:02}'.format(int(self.pd["timing"]["timecode"]["seconds"]))
            ff = '{:02}'.format(int(self.pd["timing"]["timecode"]["frames"]))
            return hh + ':' + mm + ':' + ss + ':' + ff
        else:
            return None

    # Get the PTP sample time
    def Get_sample_time(self):
        if ("timing" in self.pd) and (self.pd["timing"]["sampleTimestamp"]):
            sec = float(self.pd["timing"]["sampleTimestamp"]["seconds"])
            nsec = float(self.pd["timing"]["sampleTimestamp"]["nanoseconds"])
            asec = float(self.pd["timing"]["sampleTimestamp"]["attoseconds"])
            ctime = sec + (nsec * 0.000000001) + (asec * 0.000000000000000001)
        if self.sample_time_format == "seconds":
            return ctime
        elif self.sample_time_format == "milliseconds":
            return ctime * 1000.0
        elif self.sample_time_format == "microseconds":
            return ctime * 10000.0
        elif self.sample_time_format == "nanoseconds":
            return ctime * 100000.0
        # can't convert to timecode becuase it is PTP and has no epoch?

    # Frame rate which this timecode represents
    def Get_timecode_framerate(self):
        if ("timing" in self.pd) and (self.pd["timing"]["timecode"]["format"]["frameRate"]):
            numerator = float(self.pd["timing"]["timecode"]["format"]["frameRate"]["num"])  
            denominator = float(self.pd["timing"]["timecode"]["format"]["frameRate"]["denom"]) 
            return float(numerator / denominator)
        else:
            return None

    # Set user-preferred units for translations. 
    # Valid args: "m", "cm", "mm", "in"
    def Set_trans_units(self,unit_str):
        schema_units = self.sd["properties"]["transforms"]["items"]["items"]["properties"]["translation"]["units"]
        if self.verbose:
            print("Schema says camera translation units are {}".format(schema_units))
        print("Setting preferred translation units to: {0}".format(unit_str))
        if schema_units == "meters":
            self.trans_mult = self.Conversion_factor_from_meters(unit_str)

    # User preference for time format
    # Valid args: "timecode"
    def Set_sample_time_format(self,format_str):
        print("Setting preferred sample time format to: {0}".format(format_str))
        if (format_str in self.sample_time_formats):
            self.sample_time_format = format_str

    # Esablish a user-preference for units of focus distance.
    # Valid: "m","cm","mm","in"
    def Set_focus_distance_units(self,unit_str):
        schema_units = self.sd["properties"]["lens"]["properties"]["focusDistance"]["units"]
        if self.verbose:
            print("Schema says focus distance units are {}".format(schema_units))
        print("Setting preferred focus distance units to: {}".format(unit_str))
        if unit_str == "m":
            if schema_units == "millimeter":
                self.focus_dist_mult = .001
        elif unit_str == "cm":
            if schema_units == "millimeter":
                self.focus_dist_mult = 0.1
        elif unit_str == "mm":
            if schema_units == "millimeter":        
                self.focus_dist_mult = 1.0
        elif unit_str == "in":
            if schema_units == "millimeter":        
                self.focus_dist_mult = 1/25.4

    def Get_protocol(self):
        if (self.pd["protocol"]):
            return str(self.pd["protocol"])
        else:
            return None

    def Get_slate(self):
        if "device" in self.pd.keys():
            if self.pd["device"]["slate"]:
                return str(self.pd["device"]["slate"])
            else:
                return None
        else:
            return None

    # If present in this sample, the 'static' block has the active sensor dimensions
    def Get_sensor_dim_height(self):
        if "static" in self.pd: 
            if self.pd["static"]["camera"]["activeSensorResolution"]:
                height = int(self.pd["static"]["camera"]["activeSensorResolution"]["height"])
                return height
        return None

    # If present in this sample, the 'static' block has the active sensor dimensions
    def Get_sensor_dim_width(self):
        if "static" in self.pd: 
            if self.pd["static"]["camera"]["activeSensorResolution"]:
                width = int(self.pd["static"]["camera"]["activeSensorResolution"]["width"])
                return width
            else:
                return None
        return None

    # Find units in the schema for active sensor dimensions
    def Get_sensor_dim_units(self):
        if "static" in self.pd: 
            if self.sd["properties"]["camera"]["properties"]["activeSensorPhysicalDimensions"]:
                return str(self.sd["properties"]["camera"]["properties"]["activeSensorPhysicalDimensions"]["units"])
            else:
                return None
        else:
            return None

    # If present in this sample, the 'static' block would have the active sensor dimensions
    def Get_tracking_device_serial_number(self):
        if "static" in self.pd: 
            if self.pd["static"]["device"]["serialNumber"]:
                return str(self.pd["static"]["device"]["serialNumber"])
            else:
                return None
        return None

    def Get_focal_length(self):
        if "lens" in self.pd.keys():
            return self.pd["lens"]["focalLength"]
        else:
            return None

    def Get_focus_distance(self):
         return float(self.pd["lens"]["focusDistance"]) * self.focus_dist_mult


################
# Sample parsing below:

tsample = OTProtocol(None,None)
tsample.Set_trans_units("cm")              # end-user preferred units
tsample.Set_sample_time_format("sec")        # end-user preferred units.
tsample.Set_focus_distance_units("cm")     # end-user preferred units
print()

protocol= tsample.Get_protocol()
slate = tsample.Get_slate()
print("Detected protocol: {}".format(protocol))
print("On slate: {}".format(slate))
timecode = tsample.Get_timecode()
print("Current sample timecode: {}".format(timecode))
framerate = tsample.Get_timecode_framerate()
print("At a frame rate of: {:.5}".format(framerate))
sample_time =  tsample.Get_sample_time()
print("Sample PTP time is: {} sec".format(sample_time))
snum = tsample.Get_tracking_device_serial_number()
if snum:
    print("Tracking device serial number: {}".format(snum))
else:
    print("Unknown tracking device, wait for static sample to come in...")
posX = tsample.Get_camera_trans('X')
posY = tsample.Get_camera_trans('Y')
posZ =tsample.Get_camera_trans('Z')
print("Camera position is: ({1},{2},{3}) cm".format(timecode, posX, posY, posZ))
fl = tsample.Get_focal_length()
height = tsample.Get_sensor_dim_height()
if height:
    width = tsample.Get_sensor_dim_width()
    units = tsample.Get_sensor_dim_units()
    print("Active camera sensor height: {}, width: {} {}".format(height,width,units))
else:
    print("Unknown camera sensor, wait for static sample to come in...")
fl_units = tsample.sd["properties"]["lens"]["properties"]["focalLength"]["units"]
print("Focal length is: {} {}".format(fl,fl_units))
fd = tsample.Get_focus_distance()
print("Focus distance is: {} cm".format(fd))

File: src/test/cpp/opentrackio-parser/src/opentrackio-lib/OpenTrackIOParser.cpp
// OpenTrackIOParser.cpp
//
// Library reference code for decoding opentrackIO messages
//
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the SMTPE RIS OSVP Metadata Project
//
// nlohmann JSON library licensed under the MIT License, Copyright (c) 2013-2022 Niels Lohmann

#include "OpenTrackIOParser.h"

#include <iostream>
#include <numbers>

#include "opentrackio-cpp/OpenTrackIOProperties.h"

namespace opentrackio_parser
{
    // Shorthand namespace alias.
    namespace otiop = opentrackio::opentrackioproperties;

    std::string toString(const otiop::Protocol& protocol)
    {
        std::string version;
        for (size_t idx = 0; idx < protocol.version.size(); ++idx)
        {
            version.append(std::to_string(protocol.version[idx]));
            if (idx == protocol.version.size() - 1)
            {
                break;
            }

            version.append(".");
        }

        return version;
    }

    double ConvertFactorFromMeters(const PositionUnits unit)
    {
        switch (unit)
        {
            case PositionUnits::Meters: return 1.0;
            case PositionUnits::Centimeters: return 100.0;
            case PositionUnits::Millimeters: return 1000.0;
            case PositionUnits::Inches: return (1000.0 / 25.4);
            default: return 0.0;
        }
    }
}

OpenTrackIOSampleParser::OpenTrackIOSampleParser(const std::string& sample,
                                                 const std::string& schema,
                                                 const bool verbose) : _sampleStr(sample),
                                                                       _schemaStr(schema),
                                                                       _isVerbose(verbose)
{
}

int OpenTrackIOSampleParser::importSchema()
{
    if (!_schemaStr.empty())
    {
        try
        {
            _schemaJson = nlohmann::json::parse(_schemaStr);
        }
        catch (nlohmann::json::parse_error&)
        {
            std::cout << "Got JSONDecodeError while decoding the sample!" << std::endl;
            _schemaJson = nullptr;
        }

        if (_schemaJson.is_null())
        {
            std::cout << "Import_schema(): Failed to parse OpenTrackIO schema file." << std::endl;
        }
        else
        {
            std::cout << "Parsed the schema JSON successfully." << std::endl;
            if (_isVerbose)
            {
                std::cout << "Contents of the parsed JSON schema dict:\n" << _schemaJson.dump(4) << "\n" << std::endl;
            }
        }
    }
    else
    {
        std::cout << "ERROR: no schema provided!" << std::endl;
        return -1;
    }
    return 0;
}

bool OpenTrackIOSampleParser::parse()
{
    if (_sampleStr.empty())
    {
        std::cout << "Parse(): Error. No json text provided!" << std::endl;
        return false;
    }

    std::cout << "Parsing JSON string from sample buffer..." << std::endl;
    if (!_sample.initialise(static_cast<std::string_view>(_sampleStr)))
    {
        for (const auto& errors = _sample.getErrors();
             const auto& error : errors)
        {
            std::cout << "Error: " << error << std::endl;
        }

        return false;
    }

    if (_isVerbose)
    {
        for (const auto& warnings = _sample.getWarnings();
             const auto& warning : warnings)
        {
            std::cout << "Warning: " << warning << std::endl;
        }

        const auto json = _sample.getJson();
        std::cout << "Contents of the parsed JSON dict:\n" << json.dump(4) << "\n" << std::endl;
    }

    return true;
}

double OpenTrackIOSampleParser::getTransform(const std::string& dimension) const
{
    if (!_sample.transforms.has_value())
    {
        return 0.0;
    }

    for (const auto& transform : _sample.transforms->transforms)
    {
        if (transform.id == "Camera")
        {
            if (_isVerbose)
            {
                std::cout << "found camera, dim = " << dimension << ", mult factor: " << _transformMultiplier << std::endl;
            }

            if (dimension == "x")
            {
                return transform.translation.x * _transformMultiplier;
            }

            if (dimension == "y")
            {
                return transform.translation.y * _transformMultiplier;
            }

            if (dimension == "z")
            {
                return transform.translation.z * _transformMultiplier;
            }

            break;
        }
    }
    return 0.0;
}

double OpenTrackIOSampleParser::getRotation(const std::string& dimension) const
{
    if (!_sample.transforms.has_value())
    {
        return 0.0;
    }

    for (const auto& transform : _sample.transforms->transforms)
    {
        if (transform.id == "Camera")
        {
            if (dimension == "p")
            {
                return transform.rotation.pan * _rotationMultiplier;
            }

            if (dimension == "t")
            {
                return transform.rotation.tilt * _rotationMultiplier;
            }

            if (dimension == "r")
            {
                return transform.rotation.roll * _rotationMultiplier;
            }

            break;
        }
    }

    return 0.0;
}

std::tuple<double, double, double> OpenTrackIOSampleParser::getCameraTransform() const
{
    return std::make_tuple(getTransform("x"), getTransform("y"), getTransform("z"));
}

std::string OpenTrackIOSampleParser::getTimecode() const
{
    if (!_sample.timing.has_value() || !_sample.timing->timecode.has_value())
    {
        return "";
    }

    return std::to_string(_sample.timing->timecode->hours) + ":" +
           std::to_string(_sample.timing->timecode->minutes) + ":" +
           std::to_string(_sample.timing->timecode->seconds) + ":" +
           std::to_string(_sample.timing->timecode->frames);
}

std::string OpenTrackIOSampleParser::getSampleTime(const std::string& part) const
{
    if (!_sample.timing.has_value() || !_sample.timing->sampleTimestamp.has_value())
    {
        return "";
    }

    const uint64_t seconds = _sample.timing->sampleTimestamp->seconds;
    const uint32_t nano_seconds = _sample.timing->sampleTimestamp->nanoseconds;

    constexpr int epoch = 1'970; // PTP is since this epoch
    constexpr int spm = 60; // seconds per minute. Common knowledge, but consistency is important
    constexpr int sph = 3'600; // seconds per hour
    constexpr int spd = 86'400; // sec per day
    constexpr uint64_t spy = 31'536'000; // sec per year

    // separate into years, days, hours, min, sec
    const double y_delta = static_cast<double>(seconds) / spy; // years since epoch
    const uint32_t yr = std::floor(epoch + y_delta); // current year
    const uint32_t sty = seconds - std::floor(y_delta) * spy; // seconds elapsed this year
    const uint32_t day = std::floor(sty / spd); // current day of year
    const uint32_t std = sty - day * spd; // seconds elapsed today (since midnight)
    const uint32_t hr = std::floor(std / sph); // hours elapsed today
    const uint32_t mn = std::floor((std - hr * sph) / spm); // remainder minutes
    const uint32_t st = std - hr * sph - mn * spm; // remainder seconds

    if (part.empty())
    {
        constexpr double formattingCoefficient = 0.000000001;
        if (_sampleTimeFormat == "sec")
        {
            return std::to_string(seconds + nano_seconds * formattingCoefficient);
        }

        if (_sampleTimeFormat == "timecode")
        {
            const int frm = static_cast<int>(nano_seconds * formattingCoefficient * getSampleRate());
            return std::to_string(hr) + ":" + std::to_string(mn) + ":" + std::to_string(st) + ":" + std::to_string(frm);
        }

        if (_sampleTimeFormat == "string")
        {
            return "year:" + std::to_string(yr) + " day:" + std::to_string(day) + " hour:" + std::to_string(hr) +
                   " min:" + std::to_string(mn) + " sec:" + std::to_string(st) + " nsec:" +
                   std::to_string(nano_seconds);
        }
    }
    else
    {
        if (part == "yy")
            return std::to_string(yr);
        if (part == "dd")
            return std::to_string(day);
        if (part == "hh")
            return std::to_string(hr);
        if (part == "mm")
            return std::to_string(mn);
        if (part == "ss")
            return std::to_string(st);
        if (part == "ns")
            return std::to_string(nano_seconds);
    }
    return "";
}

double OpenTrackIOSampleParser::getSampleRate() const
{
    return _sample.timing->sampleRate->numerator /
           _sample.timing->sampleRate->denominator;
}

void OpenTrackIOSampleParser::setTranslationUnits(const opentrackio_parser::PositionUnits unit)
{
    const std::string& schema_units = _schemaJson["properties"]["transforms"]["items"]["properties"]["translation"]
            ["units"];

    if (_isVerbose)
    {
        std::cout << "Schema says camera translation units are: " << schema_units << std::endl;
        std::cout << "Setting preferred translation units to: " << toString(unit) << std::endl;
    }

    if (schema_units == "meter")
    {
        _transformMultiplier = ConvertFactorFromMeters(unit);
    }
}

void OpenTrackIOSampleParser::setRotationUnits(const opentrackio_parser::RotationUnits unit)
{
    const std::string& schemaUnit = _schemaJson["properties"]["transforms"]["items"]["properties"]["rotation"]
            ["units"];

    if (_isVerbose)
    {
        std::cout << "Schema says camera rotation units are: " << schemaUnit << std::endl;
        std::cout << "Setting preferred camera rotation units to: " << toString(unit) << std::endl;
    }

    if (constexpr auto expectedSchemaUnit = "degree";
        schemaUnit == expectedSchemaUnit)
    {
        switch (unit)
        {
            case opentrackio_parser::RotationUnits::Degrees: _rotationMultiplier = 1.0;
                break;
            case opentrackio_parser::RotationUnits::Radians: _rotationMultiplier = std::numbers::pi_v<float> / 180.f;
                break;
            default:
                assert(false && "Invalid unit for rotation");
                break;
        }

        return;
    }

    assert(false && "Invalid schema unit, expectedSchemaUnit may need updating.");
}

void OpenTrackIOSampleParser::setSampleTimeFormat(const opentrackio_parser::SampleTimeFormat format)
{
    if (_isVerbose)
    {
        const std::string& schema_units = _schemaJson["properties"]["timing"]["properties"]["sampleTimestamp"]
                ["units"];

        std::cout << "Schema says sample time units are: " << schema_units << std::endl;
        std::cout << "Setting preferred sample time format to: " << toString(format) << std::endl;
    }

    _sampleTimeFormat = toString(format);
}

void OpenTrackIOSampleParser::setFocusDistanceUnits(const opentrackio_parser::PositionUnits unit)
{
    const std::string& schemaUnit = _schemaJson["properties"]["lens"]["properties"]["focusDistance"]["units"];

    if (_isVerbose)
    {
        std::cout << "Schema says focus distance units are: " << schemaUnit << std::endl;
        std::cout << "Setting preferred focus distance units to: " << toString(unit) << std::endl;
    }

    if (constexpr auto expectedSchemaUnit = "meter";
        schemaUnit == expectedSchemaUnit)
    {
        switch (unit)
        {
            case opentrackio_parser::PositionUnits::Meters: _focusDistanceMultiplier = 0.001;
                break;
            case opentrackio_parser::PositionUnits::Centimeters: _focusDistanceMultiplier = 0.1;
                break;
            case opentrackio_parser::PositionUnits::Millimeters: _focusDistanceMultiplier = 1.0;
                break;
            case opentrackio_parser::PositionUnits::Inches: _focusDistanceMultiplier = 1.0 / 25.4;
                break;
            default:
                assert(false && "Invalid unit for focus distance");
                break;
        }

        return; // Successfully set the focus distance unit
    }

    assert(false && "Invalid schema unit, expectedSchemaUnit may need updating.");
}

std::string OpenTrackIOSampleParser::getProtocol() const
{
    if (_sample.protocol.has_value())
    {
        return _sample.protocol->name + " v" + opentrackio_parser::toString(_sample.protocol.value());
    }

    return "";
}

std::string OpenTrackIOSampleParser::getSlate() const
{
    if (_sample.tracker.has_value() && _sample.tracker->slate.has_value())
    {
        return _sample.tracker->slate.value();
    }

    return "";
}

int OpenTrackIOSampleParser::getSensoryResolutionHeight() const
{
    if (_sample.camera.has_value() && _sample.camera->activeSensorResolution.has_value())
    {
        return _sample.camera->activeSensorResolution->height;
    }

    return 0;
}

int OpenTrackIOSampleParser::getSensorResolutionWidth() const
{
    if (_sample.camera.has_value() && _sample.camera->activeSensorResolution.has_value())
    {
        return _sample.camera->activeSensorResolution->width;
    }

    return 0.0;
}

std::string OpenTrackIOSampleParser::getSensorDimensionsUnits()
{
    if (_schemaJson["properties"]["camera"].contains("activeSensorPhysicalDimensions"))
    {
        return _schemaJson["properties"]["camera"]["activeSensorPhysicalDimensions"]["units"].get<std::string>();
    }

    return "";
}

std::string OpenTrackIOSampleParser::getTrackingDeviceSerialNumber() const
{
    if (_sample.tracker.has_value() && _sample.tracker->serialNumber.has_value())
    {
        return _sample.tracker->serialNumber.value();
    }

    return "";
}

double OpenTrackIOSampleParser::getFocalLength() const
{
    if (_sample.lens.has_value() &&
        _sample.lens->pinholeFocalLength.has_value() &&
        _sample.lens->pinholeFocalLength.has_value())
    {
        return _sample.lens->pinholeFocalLength.value();
    }

    return 0.0;
}

double OpenTrackIOSampleParser::getFocusDistance() const
{
    if (_sample.lens.has_value() && _sample.lens->focusDistance.has_value())
    {
        return *_sample.lens->focusDistance * _focusDistanceMultiplier;
    }

    return 0.0;
}
